build/commands/lib/depotTools.js

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+const ActionGuard = require('./actionGuard')
+const config = require('./config')
+const fs = require('fs')
+const path = require('path')
+const Log = require('./logging')
+const util = require('./util')
+
+// This guard is used to ensure that we don't have a partial or broken checkout
+// of depot_tools.
+const depotToolsGuard = new ActionGuard(
+  path.join(
+    path.dirname(config.depotToolsDir),
+    `${path.basename(config.depotToolsDir)}_install.guard`
+  ),
+  () => {
+    if (fs.existsSync(config.depotToolsDir)) {
+      Log.status('detected incomplete depot_tools install. Cleaning up...')
+      fs.rmSync(config.depotToolsDir, { recursive: true })
+    }
+  }
+)
+
+// The .disable_auto_update file in the depot_tools directory, regardless of
+// its content, disables auto-updates. If a specific depot_tools reference is
+// enforced, we use this file for two purposes: 1) disable auto-updates; 2)
+// store the enforced reference. Note: The presence of this file also prevents
+// the auto-bootstrapping of CIPD (Chrome Infrastructure Package Deployment),
+// Python, and other dependencies.
+const disableAutoUpdateFilePath = path.join(
+  config.depotToolsDir,
+  '.disable_auto_update'
+)
+
+// Reads the enforced depot_tools ref from .disable_auto_update file. If the
+// file does not exist or the ref is not valid, returns null.
+function readEnforcedDepotToolsRef() {
+  if (!fs.existsSync(disableAutoUpdateFilePath)) {
+    return null
+  }
+  const ref = fs.readFileSync(disableAutoUpdateFilePath, 'utf8').trim()
+  const isRefValid = /^[a-f0-9]{40}$|^refs\/tags\/.*$|^origin\/.*$/.test(ref)
```

Comment:
It's not clear does the code works well with when is targeting to some branch.
Do we update depot_tools if the branch is updated? Do we need that ability? Maybe simple restrict it to sha1?

P.S. tags and sha1 should be good